src/Resolver/ValueResolver/ValueResolver.php

```diff
@@ -104,4 +110,29 @@ public function resolve(Expr $expr, Scope $scope, $unknownValuePlaceholder = nul
             return null;
         }
     }
+
+    public function resolveAsConstantType(Expr $expr, Scope $scope): ?Type
```

Comment:
Is this reallly part of ValueResolver? Imho you’re returning Type, so it should be method in TypeResolver.

Reply:
Mor me it seems closely related to vakue resolver. It is just different representation or resolved value.

Reply:
Well closely it is in this case but you just resolving Type I would like to split it

Reply:
So create ConstantValueTypeResolver in TypeaResolvers that woukd have ValueResolver as dependency?

Reply:
I prefer TypeResolver with method resolveAsConstantType